import requests

country_code = "ME"

url = f"https://api.openaq.org/v2/locations?country={country_code}"
# https://api.openaq.org/v2/locations?country=me

headers = {"Accept": "application/json"}

response = requests.get(url, headers=headers)

print(response.content)

In [3]:
import requests
import json
from IPython.display import display, JSON

country_code = "US"
API_KEY = '7b2fbb569507ae14141d3de0ff76044999ae3859816024ab9d2978c83bc3d804'

url = f"https://api.openaq.org/v2/locations?country={country_code}"
# https://api.openaq.org/v2/locations?country=me

#headers = {"Accept": "application/json"}
headers={"X-API-Key": API_KEY}

response = requests.get(url, headers=headers)

In [4]:
print(f'type response: {response}')

type response: <Response [200]>


In [5]:
data = response.json()
print(f'type: {type(data)}')

data_dumps = json.dumps(data, indent=4)
print('type: ', data_dumps)


#print(json.dumps(json_data, indent=4))

#data = json.loads(response)

#print(type(data))

#display(JSON(data_dumps))

type: <class 'dict'>
type:  {
    "meta": {
        "name": "openaq-api",
        "license": "",
        "website": "/",
        "page": 1,
        "limit": 100,
        "found": 4303
    },
    "results": [
        {
            "id": 947295,
            "city": null,
            "name": "Hamlin ES (4349)",
            "entity": null,
            "country": "US",
            "sources": null,
            "isMobile": false,
            "isAnalysis": null,
            "parameters": [
                {
                    "id": 2,
                    "unit": "\u00b5g/m\u00b3",
                    "count": 208321,
                    "average": 10.240833333333333,
                    "parameter": "pm25",
                    "lastValue": 13.78,
                    "displayName": "pm25 \u00b5g/m\u00b3",
                    "lastUpdated": "2024-09-03T09:07:11.529+00:00",
                    "parameterId": 2,
                    "firstUpdated": "2022-02-18T02:00:00+00:00",
                    

In [6]:
import pandas as pd
from pandas import json_normalize

results = json_normalize(data)

In [7]:
results.head()

results   meta.name meta.license  \
0  [{'id': 947295, 'city': None, 'name': 'Hamlin ...  openaq-api                

  meta.website  meta.page  meta.limit  meta.found  
0            /          1         100        4303

In [8]:
results_df = json_normalize(data['results'])
results_df.head()

id  city                    name entity country sources  isMobile  \
0   947295  None        Hamlin ES (4349)   None      US    None     False   
1   947163  None       Trinity ES (7219)   None      US    None     False   
2   947245  None  San Fernando MS (8358)   None      US    None     False   
3  2812820  None           East Nicolaus   None      US    None     False   
4   947238  None     Jefferson HS (8714)   None      US    None     False   

  isAnalysis                                         parameters sensorType  \
0       None  [{'id': 2, 'unit': 'µg/m³', 'count': 208321, '...       None   
1       None  [{'id': 100, 'unit': 'c', 'count': 28166, 'ave...       None   
2       None  [{'id': 128, 'unit': 'f', 'count': 32727, 'ave...       None   
3       None  [{'id': 2, 'unit': 'µg/m³', 'count': 8826, 'av...       None   
4       None  [{'id': 2, 'unit': 'µg/m³', 'count': 202660, '...       None   

                 lastUpdated               firstUpdated  measurements  \
0  2024-09-03T09:07:11+00:00  2022-02-18T01:52:16+00:00        655667   
1  2024-09-03T09:07:10+00:00  2022-02-18T01:54:11+00:00        606489   
2  2024-09-03T09:07:10+00:00  2022-02-18T01:56:39+00:00        716587   
3  2024-09-03T09:07:09+00:00  2024-04-24T16:00:00+00:00          8826   
4  2024-09-03T09:07:08+00:00  2022-02-18T01:52:52+00:00        637845   

                                         bounds  \
0  [-118.61999, 34.18983, -118.61999, 34.18983]   
1    [-118.26886, 34.0124, -118.26886, 34.0124]   
2    [-118.4366, 34.28235, -118.4366, 34.28235]   
3  [-121.54112, 38.91025, -121.54112, 38.91025]   
4  [-118.25092, 34.01106, -118.25092, 34.01106]   

                                       manufacturers  coordinates.latitude  \
0  [{'modelName': 'Clarity Sensor', 'manufacturer...              34.18983   
1  [{'modelName': 'Clarity Sensor', 'manufacturer...              34.01240   
2  [{'modelName': 'Clarity Sensor', 'manufacturer...              34.28235   
3  [{'modelName': 'Clarity Sensor', 'manufacturer...              38.91025   
4  [{'modelName': 'Clarity Sensor', 'manufacturer...              34.01106   

   coordinates.longitude  
0             -118.61999  
1             -118.26886  
2             -118.43660  
3             -121.54112  
4             -118.25092

In [9]:
results_df.shape

(100, 17)

In [10]:
results_df[['parameters']]

parameters
0   [{'id': 2, 'unit': 'µg/m³', 'count': 208321, '...
1   [{'id': 100, 'unit': 'c', 'count': 28166, 'ave...
2   [{'id': 128, 'unit': 'f', 'count': 32727, 'ave...
3   [{'id': 2, 'unit': 'µg/m³', 'count': 8826, 'av...
4   [{'id': 2, 'unit': 'µg/m³', 'count': 202660, '...
..                                                ...
95  [{'id': 2, 'unit': 'µg/m³', 'count': 57378, 'a...
96  [{'id': 19, 'unit': 'µg/m³', 'count': 178260, ...
97  [{'id': 128, 'unit': 'f', 'count': 57076, 'ave...
98  [{'id': 128, 'unit': 'f', 'count': 62295, 'ave...
99  [{'id': 2, 'unit': 'µg/m³', 'count': 8679, 'av...

[100 rows x 1 columns]

In [11]:
param_df = results_df.explode('parameters') # create repeating rows, one for each dictionary item in 'parameters'
param_df.head()

id  city              name entity country sources  isMobile isAnalysis  \
0  947295  None  Hamlin ES (4349)   None      US    None     False       None   
0  947295  None  Hamlin ES (4349)   None      US    None     False       None   
0  947295  None  Hamlin ES (4349)   None      US    None     False       None   
0  947295  None  Hamlin ES (4349)   None      US    None     False       None   
0  947295  None  Hamlin ES (4349)   None      US    None     False       None   

                                          parameters sensorType  \
0  {'id': 2, 'unit': 'µg/m³', 'count': 208321, 'a...       None   
0  {'id': 1, 'unit': 'µg/m³', 'count': 181784, 'a...       None   
0  {'id': 100, 'unit': 'c', 'count': 38079, 'aver...       None   
0  {'id': 19, 'unit': 'µg/m³', 'count': 181784, '...       None   
0  {'id': 128, 'unit': 'f', 'count': 45699, 'aver...       None   

                 lastUpdated               firstUpdated  measurements  \
0  2024-09-03T09:07:11+00:00  2022-02-18T01:52:16+00:00        655667   
0  2024-09-03T09:07:11+00:00  2022-02-18T01:52:16+00:00        655667   
0  2024-09-03T09:07:11+00:00  2022-02-18T01:52:16+00:00        655667   
0  2024-09-03T09:07:11+00:00  2022-02-18T01:52:16+00:00        655667   
0  2024-09-03T09:07:11+00:00  2022-02-18T01:52:16+00:00        655667   

                                         bounds  \
0  [-118.61999, 34.18983, -118.61999, 34.18983]   
0  [-118.61999, 34.18983, -118.61999, 34.18983]   
0  [-118.61999, 34.18983, -118.61999, 34.18983]   
0  [-118.61999, 34.18983, -118.61999, 34.18983]   
0  [-118.61999, 34.18983, -118.61999, 34.18983]   

                                       manufacturers  coordinates.latitude  \
0  [{'modelName': 'Clarity Sensor', 'manufacturer...              34.18983   
0  [{'modelName': 'Clarity Sensor', 'manufacturer...              34.18983   
0  [{'modelName': 'Clarity Sensor', 'manufacturer...              34.18983   
0  [{'modelName': 'Clarity Sensor', 'manufacturer...              34.18983   
0  [{'modelName': 'Clarity Sensor', 'manufacturer...              34.18983   

   coordinates.longitude  
0             -118.61999  
0             -118.61999  
0             -118.61999  
0             -118.61999  
0             -118.61999

In [12]:
param_normalized_df = json_normalize(param_df['parameters']) # each dictionary item on its own row
param_normalized_df.head()

id   unit   count    average    parameter  lastValue    displayName  \
0    2  µg/m³  208321  10.240833         pm25      13.78     pm25 µg/m³   
1    1  µg/m³  181784  16.172009         pm10       4.77     pm10 µg/m³   
2  100      c   38079  23.278438  temperature      23.81  temperature c   
3   19  µg/m³  181784   7.158559          pm1       1.62      pm1 µg/m³   
4  128      f   45699  16.274808  temperature      20.53  temperature f   

                     lastUpdated  parameterId                   firstUpdated  \
0  2024-09-03T09:07:11.529+00:00            2      2022-02-18T02:00:00+00:00   
1  2024-04-24T02:06:06.574+00:00            1  2022-02-18T01:52:16.232+00:00   
2  2024-04-24T00:05:03.215+00:00          100  2023-06-12T22:52:55.903+00:00   
3  2024-04-24T02:06:06.574+00:00           19  2022-02-18T01:52:16.232+00:00   
4  2024-04-24T02:06:06.574+00:00          128  2023-05-23T17:54:15.915+00:00   

  manufacturers  
0          None  
1          None  
2          None  
3          None  
4          None

In [13]:
param_normalized_df.rename(columns={'id': 'meter_id'}, inplace=True)
param_normalized_df.head()

meter_id   unit   count    average    parameter  lastValue    displayName  \
0         2  µg/m³  208321  10.240833         pm25      13.78     pm25 µg/m³   
1         1  µg/m³  181784  16.172009         pm10       4.77     pm10 µg/m³   
2       100      c   38079  23.278438  temperature      23.81  temperature c   
3        19  µg/m³  181784   7.158559          pm1       1.62      pm1 µg/m³   
4       128      f   45699  16.274808  temperature      20.53  temperature f   

                     lastUpdated  parameterId                   firstUpdated  \
0  2024-09-03T09:07:11.529+00:00            2      2022-02-18T02:00:00+00:00   
1  2024-04-24T02:06:06.574+00:00            1  2022-02-18T01:52:16.232+00:00   
2  2024-04-24T00:05:03.215+00:00          100  2023-06-12T22:52:55.903+00:00   
3  2024-04-24T02:06:06.574+00:00           19  2022-02-18T01:52:16.232+00:00   
4  2024-04-24T02:06:06.574+00:00          128  2023-05-23T17:54:15.915+00:00   

  manufacturers  
0          None  
1          None  
2          None  
3          None  
4          None

In [14]:
#df = pd.concat([param_df['id'], param_normalized_df], axis=1)

In [15]:
country_code = "ME"

url = f"https://api.openaq.org/v2/locations?country={country_code}"

# https://api.openaq.org/v2/locations?country=me

headers = {"Accept": "application/json"}

response = requests.get(url, headers=headers)

print(response.content)

b'{"message":"API KEY missing from x-api-key header"}'


## 

In [16]:
data = response.json()
print(f'type: {type(data)}')

data_dumps = json.dumps(data, indent=4)
print('type: ', data_dumps)

type: <class 'dict'>
type:  {
    "message": "API KEY missing from x-api-key header"
}


## Plaza Del Carmen (4318)

In [17]:
location_id = 4318

url = f"https://api.openaq.org/v2/locations/{location_id}"

headers = {"Accept": "application/json",
           "X-API-Key": API_KEY}

sensor_response = requests.get(url, headers=headers)

print(sensor_response.content)

b'{"meta":{"name":"openaq-api","license":"","website":"/","page":1,"limit":100,"found":1},"results":[{"id":4318,"city":"MADRID","name":"PLAZA DEL CARMEN","entity":null,"country":"ES","sources":null,"isMobile":false,"isAnalysis":null,"parameters":[{"id":3,"unit":"\xc2\xb5g/m\xc2\xb3","count":28542,"average":41.336559440064974,"parameter":"o3","lastValue":80.0,"displayName":"o3 \xc2\xb5g/m\xc2\xb3","lastUpdated":"2024-09-02T22:00:00+00:00","parameterId":3,"firstUpdated":"2016-11-18T00:00:00+00:00","manufacturers":null},{"id":5,"unit":"\xc2\xb5g/m\xc2\xb3","count":28557,"average":31.00374558303887,"parameter":"no2","lastValue":21.0,"displayName":"no2 \xc2\xb5g/m\xc2\xb3","lastUpdated":"2024-09-02T22:00:00+00:00","parameterId":5,"firstUpdated":"2016-11-18T02:00:00+00:00","manufacturers":null},{"id":4,"unit":"\xc2\xb5g/m\xc2\xb3","count":26214,"average":352.48968993432106,"parameter":"co","lastValue":500.0,"displayName":"co \xc2\xb5g/m\xc2\xb3","lastUpdated":"2024-03-11T12:00:00+00:00","par

In [18]:
sensor_data = sensor_response.json()

data_dumps = json.dumps(sensor_data, indent=4)
print('type: ', data_dumps)

type:  {
    "meta": {
        "name": "openaq-api",
        "license": "",
        "website": "/",
        "page": 1,
        "limit": 100,
        "found": 1
    },
    "results": [
        {
            "id": 4318,
            "city": "MADRID",
            "name": "PLAZA DEL CARMEN",
            "entity": null,
            "country": "ES",
            "sources": null,
            "isMobile": false,
            "isAnalysis": null,
            "parameters": [
                {
                    "id": 3,
                    "unit": "\u00b5g/m\u00b3",
                    "count": 28542,
                    "average": 41.336559440064974,
                    "parameter": "o3",
                    "lastValue": 80.0,
                    "displayName": "o3 \u00b5g/m\u00b3",
                    "lastUpdated": "2024-09-02T22:00:00+00:00",
                    "parameterId": 3,
                    "firstUpdated": "2016-11-18T00:00:00+00:00",
                    "manufacturers": null
          

_A note on the output:_ The `results` item contains data about the sensor and its location, as well as a dictionary of `parameters` that contian the actual sensor readings. We want to duplicate the data related to the sensor across separate lines for seaparate measurements.

Get output as df

In [19]:
sensor_df = json_normalize(sensor_data['results'])
sensor_df.head()

id    city              name entity country sources  isMobile isAnalysis  \
0  4318  MADRID  PLAZA DEL CARMEN   None      ES    None     False       None   

                                          parameters sensorType  \
0  [{'id': 3, 'unit': 'µg/m³', 'count': 28542, 'a...       None   

                 lastUpdated               firstUpdated  measurements  \
0  2024-09-02T22:00:00+00:00  2016-11-18T00:00:00+00:00        116042   

                                              bounds  \
0  [-3.703333329700339, 40.41916666016447, -3.703...   

                                       manufacturers  coordinates.latitude  \
0  [{'modelName': 'Government Monitor', 'manufact...             40.419167   

   coordinates.longitude  
0              -3.703333

Recreate with one row per parameter

In [20]:
sensor_df = sensor_df.explode('parameters', ignore_index=True) # create repeating rows, one for each dictionary item in 'parameters'
sensor_df

id    city              name entity country sources  isMobile isAnalysis  \
0  4318  MADRID  PLAZA DEL CARMEN   None      ES    None     False       None   
1  4318  MADRID  PLAZA DEL CARMEN   None      ES    None     False       None   
2  4318  MADRID  PLAZA DEL CARMEN   None      ES    None     False       None   
3  4318  MADRID  PLAZA DEL CARMEN   None      ES    None     False       None   
4  4318  MADRID  PLAZA DEL CARMEN   None      ES    None     False       None   

                                          parameters sensorType  \
0  {'id': 3, 'unit': 'µg/m³', 'count': 28542, 'av...       None   
1  {'id': 5, 'unit': 'µg/m³', 'count': 28557, 'av...       None   
2  {'id': 4, 'unit': 'µg/m³', 'count': 26214, 'av...       None   
3  {'id': 19843, 'unit': 'µg/m³', 'count': 4426, ...       None   
4  {'id': 6, 'unit': 'µg/m³', 'count': 28303, 'av...       None   

                 lastUpdated               firstUpdated  measurements  \
0  2024-09-02T22:00:00+00:00  2016-11-18T00:00:00+00:00        116042   
1  2024-09-02T22:00:00+00:00  2016-11-18T00:00:00+00:00        116042   
2  2024-09-02T22:00:00+00:00  2016-11-18T00:00:00+00:00        116042   
3  2024-09-02T22:00:00+00:00  2016-11-18T00:00:00+00:00        116042   
4  2024-09-02T22:00:00+00:00  2016-11-18T00:00:00+00:00        116042   

                                              bounds  \
0  [-3.703333329700339, 40.41916666016447, -3.703...   
1  [-3.703333329700339, 40.41916666016447, -3.703...   
2  [-3.703333329700339, 40.41916666016447, -3.703...   
3  [-3.703333329700339, 40.41916666016447, -3.703...   
4  [-3.703333329700339, 40.41916666016447, -3.703...   

                                       manufacturers  coordinates.latitude  \
0  [{'modelName': 'Government Monitor', 'manufact...             40.419167   
1  [{'modelName': 'Government Monitor', 'manufact...             40.419167   
2  [{'modelName': 'Government Monitor', 'manufact...             40.419167   
3  [{'modelName': 'Government Monitor', 'manufact...             40.419167   
4  [{'modelName': 'Government Monitor', 'manufact...             40.419167   

   coordinates.longitude  
0              -3.703333  
1              -3.703333  
2              -3.703333  
3              -3.703333  
4              -3.703333

In [21]:
param_df = json_normalize(sensor_df['parameters']) # extract parameters to separate dataframe 
param_df

id   unit  count     average parameter  lastValue displayName  \
0      3  µg/m³  28542   41.336559        o3       80.0    o3 µg/m³   
1      5  µg/m³  28557   31.003746       no2       21.0   no2 µg/m³   
2      4  µg/m³  26214  352.489690        co      500.0    co µg/m³   
3  19843  µg/m³   4426   12.596701        no        3.0    no µg/m³   
4      6  µg/m³  28303    5.231899       so2        3.0   so2 µg/m³   

                 lastUpdated  parameterId               firstUpdated  \
0  2024-09-02T22:00:00+00:00            3  2016-11-18T00:00:00+00:00   
1  2024-09-02T22:00:00+00:00            5  2016-11-18T02:00:00+00:00   
2  2024-03-11T12:00:00+00:00            4  2016-11-18T00:00:00+00:00   
3  2024-03-05T00:00:00+00:00        19843  2023-03-29T11:00:00+00:00   
4  2024-09-02T22:00:00+00:00            6  2016-11-18T00:00:00+00:00   

  manufacturers  
0          None  
1          None  
2          None  
3          None  
4          None

Combine dataframes to include sensor data and parameter readings

In [22]:
readings_df = pd.concat([sensor_df, param_df], axis=1)
readings_df

id    city              name entity country sources  isMobile isAnalysis  \
0  4318  MADRID  PLAZA DEL CARMEN   None      ES    None     False       None   
1  4318  MADRID  PLAZA DEL CARMEN   None      ES    None     False       None   
2  4318  MADRID  PLAZA DEL CARMEN   None      ES    None     False       None   
3  4318  MADRID  PLAZA DEL CARMEN   None      ES    None     False       None   
4  4318  MADRID  PLAZA DEL CARMEN   None      ES    None     False       None   

                                          parameters sensorType  ...   unit  \
0  {'id': 3, 'unit': 'µg/m³', 'count': 28542, 'av...       None  ...  µg/m³   
1  {'id': 5, 'unit': 'µg/m³', 'count': 28557, 'av...       None  ...  µg/m³   
2  {'id': 4, 'unit': 'µg/m³', 'count': 26214, 'av...       None  ...  µg/m³   
3  {'id': 19843, 'unit': 'µg/m³', 'count': 4426, ...       None  ...  µg/m³   
4  {'id': 6, 'unit': 'µg/m³', 'count': 28303, 'av...       None  ...  µg/m³   

   count     average parameter lastValue  displayName  \
0  28542   41.336559        o3      80.0     o3 µg/m³   
1  28557   31.003746       no2      21.0    no2 µg/m³   
2  26214  352.489690        co     500.0     co µg/m³   
3   4426   12.596701        no       3.0     no µg/m³   
4  28303    5.231899       so2       3.0    so2 µg/m³   

                 lastUpdated  parameterId               firstUpdated  \
0  2024-09-02T22:00:00+00:00            3  2016-11-18T00:00:00+00:00   
1  2024-09-02T22:00:00+00:00            5  2016-11-18T02:00:00+00:00   
2  2024-03-11T12:00:00+00:00            4  2016-11-18T00:00:00+00:00   
3  2024-03-05T00:00:00+00:00        19843  2023-03-29T11:00:00+00:00   
4  2024-09-02T22:00:00+00:00            6  2016-11-18T00:00:00+00:00   

   manufacturers  
0           None  
1           None  
2           None  
3           None  
4           None  

[5 rows x 28 columns]

Drop `parameters` column

In [23]:
readings_df.drop(columns='parameters')
all_readings_df = readings_df
all_readings_df

id    city              name entity country sources  isMobile isAnalysis  \
0  4318  MADRID  PLAZA DEL CARMEN   None      ES    None     False       None   
1  4318  MADRID  PLAZA DEL CARMEN   None      ES    None     False       None   
2  4318  MADRID  PLAZA DEL CARMEN   None      ES    None     False       None   
3  4318  MADRID  PLAZA DEL CARMEN   None      ES    None     False       None   
4  4318  MADRID  PLAZA DEL CARMEN   None      ES    None     False       None   

                                          parameters sensorType  ...   unit  \
0  {'id': 3, 'unit': 'µg/m³', 'count': 28542, 'av...       None  ...  µg/m³   
1  {'id': 5, 'unit': 'µg/m³', 'count': 28557, 'av...       None  ...  µg/m³   
2  {'id': 4, 'unit': 'µg/m³', 'count': 26214, 'av...       None  ...  µg/m³   
3  {'id': 19843, 'unit': 'µg/m³', 'count': 4426, ...       None  ...  µg/m³   
4  {'id': 6, 'unit': 'µg/m³', 'count': 28303, 'av...       None  ...  µg/m³   

   count     average parameter lastValue  displayName  \
0  28542   41.336559        o3      80.0     o3 µg/m³   
1  28557   31.003746       no2      21.0    no2 µg/m³   
2  26214  352.489690        co     500.0     co µg/m³   
3   4426   12.596701        no       3.0     no µg/m³   
4  28303    5.231899       so2       3.0    so2 µg/m³   

                 lastUpdated  parameterId               firstUpdated  \
0  2024-09-02T22:00:00+00:00            3  2016-11-18T00:00:00+00:00   
1  2024-09-02T22:00:00+00:00            5  2016-11-18T02:00:00+00:00   
2  2024-03-11T12:00:00+00:00            4  2016-11-18T00:00:00+00:00   
3  2024-03-05T00:00:00+00:00        19843  2023-03-29T11:00:00+00:00   
4  2024-09-02T22:00:00+00:00            6  2016-11-18T00:00:00+00:00   

   manufacturers  
0           None  
1           None  
2           None  
3           None  
4           None  

[5 rows x 28 columns]

In [68]:
url = "https://api.openaq.org/v2/parameters?limit=100&page=1&offset=0&sort=asc&order_by=id"

response = requests.get(url, headers=headers).json()


#print(json.dumps(response, indent=4))

df = pd.DataFrame(response['results'])
df = df.sort_values(by=['name'])
pd.set_option('display.max_rows', None)
df

id              name           displayName  \
28     37      ambient_temp       Temperature (C)   
58  19855                as                    As   
48  19841                bc                    BC   
13     17                bc                    BC   
10     11                bc                    BC   
59  19856                ca                    Ca   
21     28               ch4                   CH₄   
54  19851                cl                    Cl   
34    102                co                    CO   
3       4                co               CO mass   
7       8                co                    CO   
11     14               co2                   CO2   
15     21               co2                   CO₂   
52  19849                ec                    EC   
60  19857                fe                    Fe   
43    134          humidity                     H   
55  19852                 k                     K   
61  19858                ni                    Ni   
26     35                no                    NO   
18     24                no                    NO   
49  19843                no               NO mass   
6       7               no2                   NO₂   
4       5               no2              NO₂ mass   
12     15               no2                   NO₂   
56  19853               no3                   NO3   
47  19840               nox                   NOx   
17     23               nox                   NOX   
20     27               nox              NOx mass   
9      10                o3                    O₃   
2       3                o3               O₃ mass   
23     32                o3                    O₃   
53  19850                oc                    OC   
46    676             ozone                    O₃   
57  19854                pb                    Pb   
27     36                pm                    PM   
14     19               pm1                   PM1   
0       1              pm10                  PM10   
40    131             pm100                 PM100   
63  19860              pm25                  None   
1       2              pm25                 PM2.5   
30     97          pm25-old                  None   
50  19844               pm4                 PM4.0   
22     29                pn                  None   
41    132          pressure              Pressure   
29     95          pressure  Atmospheric pressure   
31     98  relativehumidity                    RH   
19     25                rh                    RH   
5       6               so2              SO₂ mass   
8       9               so2                   SO₂   
33    101               so2                   SO₂   
51  19848               so4                   SO4   
32    100       temperature                     T   
37    128       temperature       Temperature (F)   
24     33               ufp             UFP count   
35    125             um003           PM0.3 count   
42    133             um005                 PM0.5   
36    126             um010             PM1 count   
39    130             um025           PM2.5 count   
38    129             um050           PM5.0 count   
44    135             um100            PM10 count   
62  19859                 v                     V   
45    150               voc                   VOC   
16     22    wind_direction        Wind direction   
25     34        wind_speed            Wind speed   

                                          description  preferredUnit  
28                                    Temperature (C)          deg_c  
58                                            Arsenic            ppb  
48                                                 BC            ppm  
13                                                 BC          ng/m3  
10                    Black Carbon mass concentration          µg/m³  
59                                            Calcium            ppb  
21                              Methane concentration           

**pm2.5**: 2 ug/m3 (97 is old, 19860 measures ppm instead)

**pm10**: 1 ug/m3

**SO2**: 6 ug/m3 (9 measures ppm, 101  measures ppb)

**NO**: 19843 ug/m3 (24 measures ppb, 35 measures ppm)

**O3**: 3 ug/m3 (10 measures ppm, 32 measures ppb, 676 also ppb)

### 3223

In [87]:
#meter ID, location name, lat, lon, day, time, pm2.5, pm10, so2, no, o3

location_id = 3223

url = f'https://api.openaq.org/v2/locations/{location_id}'

sensor_response = requests.get(url, headers=headers)


sensor_data = sensor_response.json()

sensor_dict = sensor_data.get('results', [])[0]
params = sensor_dict.get('parameters')



print(type(sensor_dict))
print(type(params))

print(sensor_dict)
print(params)

print(params[1]['parameterId'])

#row



<class 'dict'>
<class 'list'>
{'id': 3223, 'city': 'MADRID', 'name': 'PLAZA DE ESPAÑA', 'entity': None, 'country': 'ES', 'sources': None, 'isMobile': False, 'isAnalysis': None, 'parameters': [{'id': 6, 'unit': 'µg/m³', 'count': 27374, 'average': 3.474650639725674, 'parameter': 'so2', 'lastValue': 2.0, 'displayName': 'so2 µg/m³', 'lastUpdated': '2024-09-02T22:00:00+00:00', 'parameterId': 6, 'firstUpdated': '2016-11-18T00:00:00+00:00', 'manufacturers': None}, {'id': 5, 'unit': 'µg/m³', 'count': 27301, 'average': 29.297914818101155, 'parameter': 'no2', 'lastValue': 17.0, 'displayName': 'no2 µg/m³', 'lastUpdated': '2024-09-02T22:00:00+00:00', 'parameterId': 5, 'firstUpdated': '2016-11-18T00:00:00+00:00', 'manufacturers': None}, {'id': 4, 'unit': 'µg/m³', 'count': 24873, 'average': 331.7543365396225, 'parameter': 'co', 'lastValue': 300.0, 'displayName': 'co µg/m³', 'lastUpdated': '2024-03-11T12:00:00+00:00', 'parameterId': 4, 'firstUpdated': '2016-11-18T00:00:00+00:00', 'manufacturers': Non

In [101]:
row = {'id': sensor_dict.get('id'),
       'location': sensor_dict.get('name'),
       'lat': sensor_dict.get('coordinates', {}).get('latitude'),
       'lon': sensor_dict.get('coordinates', {}).get('longitude'),
       'day': sensor_dict.get('lastUpdated'),
       'time': sensor_dict.get('lastUpdated'),
       'pm25': '',
       'pm10': '',
       's02': '',
       'no': '',
       'o3': ''
       }

print(type(row))

for param in sensor_dict.get('parameters', []):
       if param.get("parameterId") in [2, 1, 6, 19843, 3]:
              row.update({
              param['parameter']: param.get('lastValue')
       })

<class 'dict'>


In [103]:
df = pd.DataFrame([row])
df

df.to_csv('sensor_data.csv', index=False)

In [75]:
sensor_data = sensor_response.json()
sensor_dict = sensor_data.get('results', [])[0]

print(sensor_results)




{'id': 3223, 'city': 'MADRID', 'name': 'PLAZA DE ESPAÑA', 'entity': None, 'country': 'ES', 'sources': None, 'isMobile': False, 'isAnalysis': None, 'parameters': [{'id': 6, 'unit': 'µg/m³', 'count': 27374, 'average': 3.474650639725674, 'parameter': 'so2', 'lastValue': 2.0, 'displayName': 'so2 µg/m³', 'lastUpdated': '2024-09-02T22:00:00+00:00', 'parameterId': 6, 'firstUpdated': '2016-11-18T00:00:00+00:00', 'manufacturers': None}, {'id': 5, 'unit': 'µg/m³', 'count': 27301, 'average': 29.297914818101155, 'parameter': 'no2', 'lastValue': 17.0, 'displayName': 'no2 µg/m³', 'lastUpdated': '2024-09-02T22:00:00+00:00', 'parameterId': 5, 'firstUpdated': '2016-11-18T00:00:00+00:00', 'manufacturers': None}, {'id': 4, 'unit': 'µg/m³', 'count': 24873, 'average': 331.7543365396225, 'parameter': 'co', 'lastValue': 300.0, 'displayName': 'co µg/m³', 'lastUpdated': '2024-03-11T12:00:00+00:00', 'parameterId': 4, 'firstUpdated': '2016-11-18T00:00:00+00:00', 'manufacturers': None}, {'id': 19843, 'unit': 'µg/

In [112]:
url = f'https://api.openaq.org/v2/locations/{location_id}'

sensor_response = requests.get(url, headers=headers)

sensor_data = sensor_response.json()

sensor_dict = sensor_data.get('results', [])[0]

row = {'id': sensor_dict.get('id'),
       'location': sensor_dict.get('name'),
       'lat': sensor_dict.get('coordinates', {}).get('latitude'),
       'lon': sensor_dict.get('coordinates', {}).get('longitude'),
       'day': sensor_dict.get('lastUpdated'),
       'time': sensor_dict.get('lastUpdated'),
       'pm25': '',
       'pm10': '',
       's02': '',
       'no': '',
       'o3': ''
       }

for param in sensor_dict.get('parameters', []):
       if param.get("parameterId") in [2, 1, 6, 19843, 3]:
              row.update({
              param['parameter']: param.get('lastValue')})

df = pd.DataFrame([row])

file_name = 'sensor_data.csv'

df.to_csv(file_name, mode='a', header=not pd.io.common.file_exists(file_name), index=False)

In [116]:
sensor_dict.get('lastUpdated')[:10]

'2024-09-02'

In [117]:
sensor_dict.get('lastUpdated')[11:16]

'22:00'

### Ok, nice. Let's say I wanted all of Madrid...

In [119]:
url = "https://api.openaq.org/v2/cities?limit=100&page=1&offset=0&sort=asc&city=Madrid&order_by=city"

city_response = requests.get(url, headers=headers)

print(json_dumps(city_response.text.json(), indent=4)

SyntaxError: incomplete input (314231579.py, line 5)

In [26]:
madrid_data = city_response.json()

data_dumps = json.dumps(madrid_data, indent=4)
print('type: ', data_dumps)

type:  {
    "meta": {
        "name": "openaq-api",
        "license": "",
        "website": "/",
        "page": 1,
        "limit": 100,
        "found": 1
    },
    "results": [
        {
            "country": "ES",
            "city": "Madrid",
            "count": 2226261,
            "locations": 22,
            "firstUpdated": "2016-11-18 02:00:00+00",
            "lastUpdated": "2024-01-29 20:00:00+00",
            "parameters": [
                "co",
                "no",
                "no2",
                "o3",
                "pm10",
                "pm25",
                "so2"
            ]
        }
    ]
}


Interesting. That gave me a count of the sensors that were in Madrid, and what Measurements that were available, but not the IDs of the sensors.

In [27]:
lat = '40.42188'
lon = '-3.70079'
radius = '5000'
start = '2024-08-30T00%3A00%3A00Z'
stop = '2024-08-31T00%3A00%3A00Z'
url = f'https://api.openaq.org/v2/measurements?ldate_from={start}&date_to={stop}&imit=1000&page=9&offset=0&sort=desc&coordinates={lat}%2C{lon}1&radius={radius}&order_by=datetime'

response = requests.get(url, headers=headers)

data = response.json()

data_dumps = json.dumps(data, indent=4)

print('type: ', data_dumps)

type:  {
    "meta": {
        "name": "openaq-api",
        "license": "",
        "website": "/",
        "page": 9,
        "limit": 100,
        "found": 0
    },
    "results": []
}


In [28]:
lalo_df = json_normalize(data['results'])
lalo_df

Empty DataFrame
Columns: []
Index: []

In [29]:
lalo_df.shape

(0, 0)

In [30]:
lalo_df['locationId'].unique()

KeyError: 'locationId'

In [118]:
url = "https://api.openaq.org/v2/locations?limit=100&page=1&offset=0&sort=desc&coordinates=40.42188%2C-3.70079&radius=5000&order_by=lastUpdated&dump_raw=false"

response = requests.get(url, headers=headers)

data = response.json()

data_dumps = json.dumps(data, indent=4)

print('type: ', data_dumps)

type:  {
    "meta": {
        "name": "openaq-api",
        "license": "",
        "website": "/",
        "page": 1,
        "limit": 100,
        "found": 15
    },
    "results": [
        {
            "id": 4274,
            "city": "MADRID",
            "name": "CASA DE CAMPO",
            "entity": null,
            "country": "ES",
            "sources": null,
            "isMobile": false,
            "isAnalysis": null,
            "parameters": [
                {
                    "id": 2,
                    "unit": "\u00b5g/m\u00b3",
                    "count": 29197,
                    "average": 6.9691170374464555,
                    "parameter": "pm25",
                    "lastValue": 7.0,
                    "displayName": "pm25 \u00b5g/m\u00b3",
                    "lastUpdated": "2024-09-02T22:00:00+00:00",
                    "parameterId": 2,
                    "firstUpdated": "2016-11-21T20:00:00+00:00",
                    "manufacturers": null
         

In [ ]:
madrid = []
list = data['results']

print(list)
for i in data:
    print(list[i])

[{'id': 4327, 'city': 'MADRID', 'name': 'FAROLILLO', 'entity': None, 'country': 'ES', 'sources': None, 'isMobile': False, 'isAnalysis': None, 'parameters': [{'id': 19843, 'unit': 'µg/m³', 'count': 5041, 'average': 9.614560603054949, 'parameter': 'no', 'lastValue': 1.0, 'displayName': 'no µg/m³', 'lastUpdated': '2024-03-05T00:00:00+00:00', 'parameterId': 19843, 'firstUpdated': '2023-05-07T02:00:00+00:00', 'manufacturers': None}, {'id': 3, 'unit': 'µg/m³', 'count': 9450, 'average': 57.594884577869294, 'parameter': 'o3', 'lastValue': 77.0, 'displayName': 'o3 µg/m³', 'lastUpdated': '2024-09-01T22:00:00+00:00', 'parameterId': 3, 'firstUpdated': '2023-05-06T02:00:00+00:00', 'manufacturers': None}, {'id': 1, 'unit': 'µg/m³', 'count': 9451, 'average': 20.488568642323113, 'parameter': 'pm10', 'lastValue': 53.0, 'displayName': 'pm10 µg/m³', 'lastUpdated': '2024-09-01T22:00:00+00:00', 'parameterId': 1, 'firstUpdated': '2023-05-06T02:00:00+00:00', 'manufacturers': None}, {'id': 5, 'unit': 'µg/m³',

TypeError: list indices must be integers or slices, not str

Get sensor data by lat lon + 10km radius

In [162]:
lat = '40.42188'
lon = '-3.70079'
radius = '10000'

url = "https://api.openaq.org/v2/locations?limit=100&page=1&offset=0&sort=desc&coordinates=40.42188%2C-3.70079&radius=10000&order_by=lastUpdated&dump_raw=false"

sensors_response = requests.get(url, headers=headers)

sensors_data = sensors_response.json()

data_dumps = json.dumps(sensors_data, indent=4)

print('type: ', data_dumps)

type:  {
    "meta": {
        "name": "openaq-api",
        "license": "",
        "website": "/",
        "page": 1,
        "limit": 100,
        "found": 26
    },
    "results": [
        {
            "id": 4331,
            "city": "MADRID",
            "name": "MENDEZ ALVARO",
            "entity": null,
            "country": "ES",
            "sources": null,
            "isMobile": false,
            "isAnalysis": null,
            "parameters": [
                {
                    "id": 19843,
                    "unit": "\u00b5g/m\u00b3",
                    "count": 5072,
                    "average": 10.393138801261829,
                    "parameter": "no",
                    "lastValue": 1.0,
                    "displayName": "no \u00b5g/m\u00b3",
                    "lastUpdated": "2024-03-05T00:00:00+00:00",
                    "parameterId": 19843,
                    "firstUpdated": "2023-05-07T02:00:00+00:00",
                    "manufacturers": null
      

lets look at that like a dataframe

In [163]:
sensors_df = json_normalize(sensors_data['results'])
sensors_df = sensors_df.drop(columns=['entity', 'sources', 'isMobile', 'isAnalysis', 'sensorType', 'measurements', 'bounds', 'manufacturers'])

In [164]:
#sensor_df['params'] = sensor_df['parameters']['parameter']
print(type(sensors_df['parameters']))

<class 'pandas.core.series.Series'>


`parameters` contain list of dicts containing the specific parameters each sensor measures. We want to extract the type of sensor (`id` or `parameterId` - is there a difference?) maybe also the name (`parameter`) for readability.

In [165]:
sensors_df['parameterIds'] = sensors_df['parameters'].apply(lambda x: [param_dict['id'] for param_dict in x])
sensors_df

id                city                           name country  \
0      4331              MADRID                  MENDEZ ALVARO      ES   
1      4323              MADRID               ESCUELAS AGUIRRE      ES   
2      4287              MADRID                    TRES OLIVOS      ES   
3      4258              MADRID               BARRIO DEL PILAR      ES   
4      4312              MADRID           ENSANCHE DE VALLECAS      ES   
5      4314              MADRID                         RETIRO      ES   
6      4328              MADRID                 PLAZA ELÍPTICA      ES   
7      3295              MADRID                  RAMÓN Y CAJAL      ES   
8      4285              MADRID  CUATRO CAMINOS-PABLO IGLESIAS      ES   
9      3223              MADRID                PLAZA DE ESPAÑA      ES   
10  2162414  POZUELO DE ALARCÓN             POZUELO DE ALARCÓN      ES   
11     4318              MADRID               PLAZA DEL CARMEN      ES   
12     4329              MADRID                      MORATALAZ      ES   
13     4275              MADRID                     CASTELLANA      ES   
14     4327              MADRID                      FAROLILLO      ES   
15     4274              MADRID                  CASA DE CAMPO      ES   
16     4264              MADRID                     VILLAVERDE      ES   
17     4296              MADRID                    SANCHINARRO      ES   
18  2162441  POZUELO DE ALARCÓN             POZUELO DE ALARCÓN      ES   
19     4342              MADRID                  JUAN CARLOS I      ES   
20     4320              MADRID           PLAZA CASTILLA-CANAL      ES   
21     4317              MADRID             PUENTE DE VALLECAS      ES   
22     4249              MADRID                   ARTURO SORIA      ES   
23  2826460                None             Invernadero SinCO2      ES   
24  2826464                None             Invernadero ConCO2      ES   
25     3265              Madrid                        ES1567A      ES   

                                           parameters  \
0   [{'id': 19843, 'unit': 'µg/m³', 'count': 5072,...   
1   [{'id': 6, 'unit': 'µg/m³', 'count': 29315, 'a...   
2   [{'id': 3, 'unit': 'µg/m³', 'count': 30319, 'a...   
3   [{'id': 3, 'unit': 'µg/m³', 'count': 30179, 'a...   
4   [{'id': 3, 'unit': 'µg/m³', 'count': 30353, 'a...   
5   [{'id': 5, 'unit': 'µg/m³', 'count': 30223, 'a...   
6   [{'id': 4, 'unit': 'µg/m³', 'count': 27774, 'a...   
7   [{'id': 19843, 'unit': 'µg/m³', 'count': 5930,...   
8   [{'id': 5, 'unit': 'µg/m³', 'count': 30144, 'a...   
9   [{'id': 5, 'unit': 'µg/m³', 'count': 27303, 'a...   
10  [{'id': 2, 'unit': 'µg/m³', 'count': 1651, 'av...   
11  [{'id': 5, 'unit': 'µg/m³', 'count': 28559, 'a...   
12  [{'id': 4, 'unit': 'µg/m³', 'count': 4863, 'av...   
13  [{'id': 1, 'unit': 'µg/m³', 'count': 29777, 'a...   
14  [{'id': 1, 'unit': 'µg/m³', 'count': 9455, 'av...   
15  [{'id': 19843, 'unit': 'µg/m³', 'count': 5777,...   
16  [{'id': 5, 'unit': 'µg/m³', 'count': 30258, 'a...   
17  [{'id': 2, 'unit': 'µg/m³', 'count': 13311, 'a...   
18  [{'id': 19843, 'unit': 'µg/m³', 'count': 113, ...   
19  [{'id': 19843, 'unit': 'µg/m³', 'count': 5907,...   
20  [{'id': 5, 'unit': 'µg/m³', 'count': 30425, 'a...   
21  [{'id': 19843, 'unit': 'µg/m³', 'count': 5934,...   
22  [{'id': 3, 'unit': 'µg/m³', 'count': 30211, 'a...   
23  [{'id': 19, 'unit': 'µg/m³', 'count': 1249, 'a...   
24  [{'id': 19, 'unit': 'µg/m³', 'count': 885, 'av...   
25  [{'id': 19843, 'unit': 'µg/m³', 'count': 5770,...   

                  lastUpdated               firstUpdated  \
0   2024-09-03T22:00:00+00:00  2023-05-06T02:00:00+00:00   
1   2024-09-03T22:00:00+00:00  2016-11-18T00:00:00+00:00   
2   2024-09-03T22:00:00+00:00  2016-11-18T00:00:00+00:00   
3   2024-09-03T22:00:00+00:00  2016-11-18T00:00:00+00:00   
4   2024-09-03T22:00:00+00:00  2016-11-18T00:00:00+00:00   
5   2024-09-03T22:00:00+00:00  2016-11-18T00:00:00+00:00   
6   2024-09-03T22:00:00+00:00  2016-11-18T00:00:00

In [166]:
# 2, 1, 6, 19843, 3
sensors_df['has_pm25'] = sensors_df['parameterIds'].apply(lambda x: 1 if 2 in x else 0)
sensors_df['has_pm10'] = sensors_df['parameterIds'].apply(lambda x: 1 if 1 in x else 0)
sensors_df['has_so2'] = sensors_df['parameterIds'].apply(lambda x: 1 if 6 in x else 0)
sensors_df['has_no'] = sensors_df['parameterIds'].apply(lambda x: 1 if 19843 in x else 0)
sensors_df['has_o3'] = sensors_df['parameterIds'].apply(lambda x: 1 if 3 in x else 0)
sensors_df['sensorCount'] = sensors_df['has_pm25'] + sensors_df['has_pm10'] + sensors_df['has_so2'] + sensors_df['has_no'] + sensors_df['has_o3']
sensors_df

id                city                           name country  \
0      4331              MADRID                  MENDEZ ALVARO      ES   
1      4323              MADRID               ESCUELAS AGUIRRE      ES   
2      4287              MADRID                    TRES OLIVOS      ES   
3      4258              MADRID               BARRIO DEL PILAR      ES   
4      4312              MADRID           ENSANCHE DE VALLECAS      ES   
5      4314              MADRID                         RETIRO      ES   
6      4328              MADRID                 PLAZA ELÍPTICA      ES   
7      3295              MADRID                  RAMÓN Y CAJAL      ES   
8      4285              MADRID  CUATRO CAMINOS-PABLO IGLESIAS      ES   
9      3223              MADRID                PLAZA DE ESPAÑA      ES   
10  2162414  POZUELO DE ALARCÓN             POZUELO DE ALARCÓN      ES   
11     4318              MADRID               PLAZA DEL CARMEN      ES   
12     4329              MADRID                      MORATALAZ      ES   
13     4275              MADRID                     CASTELLANA      ES   
14     4327              MADRID                      FAROLILLO      ES   
15     4274              MADRID                  CASA DE CAMPO      ES   
16     4264              MADRID                     VILLAVERDE      ES   
17     4296              MADRID                    SANCHINARRO      ES   
18  2162441  POZUELO DE ALARCÓN             POZUELO DE ALARCÓN      ES   
19     4342              MADRID                  JUAN CARLOS I      ES   
20     4320              MADRID           PLAZA CASTILLA-CANAL      ES   
21     4317              MADRID             PUENTE DE VALLECAS      ES   
22     4249              MADRID                   ARTURO SORIA      ES   
23  2826460                None             Invernadero SinCO2      ES   
24  2826464                None             Invernadero ConCO2      ES   
25     3265              Madrid                        ES1567A      ES   

                                           parameters  \
0   [{'id': 19843, 'unit': 'µg/m³', 'count': 5072,...   
1   [{'id': 6, 'unit': 'µg/m³', 'count': 29315, 'a...   
2   [{'id': 3, 'unit': 'µg/m³', 'count': 30319, 'a...   
3   [{'id': 3, 'unit': 'µg/m³', 'count': 30179, 'a...   
4   [{'id': 3, 'unit': 'µg/m³', 'count': 30353, 'a...   
5   [{'id': 5, 'unit': 'µg/m³', 'count': 30223, 'a...   
6   [{'id': 4, 'unit': 'µg/m³', 'count': 27774, 'a...   
7   [{'id': 19843, 'unit': 'µg/m³', 'count': 5930,...   
8   [{'id': 5, 'unit': 'µg/m³', 'count': 30144, 'a...   
9   [{'id': 5, 'unit': 'µg/m³', 'count': 27303, 'a...   
10  [{'id': 2, 'unit': 'µg/m³', 'count': 1651, 'av...   
11  [{'id': 5, 'unit': 'µg/m³', 'count': 28559, 'a...   
12  [{'id': 4, 'unit': 'µg/m³', 'count': 4863, 'av...   
13  [{'id': 1, 'unit': 'µg/m³', 'count': 29777, 'a...   
14  [{'id': 1, 'unit': 'µg/m³', 'count': 9455, 'av...   
15  [{'id': 19843, 'unit': 'µg/m³', 'count': 5777,...   
16  [{'id': 5, 'unit': 'µg/m³', 'count': 30258, 'a...   
17  [{'id': 2, 'unit': 'µg/m³', 'count': 13311, 'a...   
18  [{'id': 19843, 'unit': 'µg/m³', 'count': 113, ...   
19  [{'id': 19843, 'unit': 'µg/m³', 'count': 5907,...   
20  [{'id': 5, 'unit': 'µg/m³', 'count': 30425, 'a...   
21  [{'id': 19843, 'unit': 'µg/m³', 'count': 5934,...   
22  [{'id': 3, 'unit': 'µg/m³', 'count': 30211, 'a...   
23  [{'id': 19, 'unit': 'µg/m³', 'count': 1249, 'a...   
24  [{'id': 19, 'unit': 'µg/m³', 'count': 885, 'av...   
25  [{'id': 19843, 'unit': 'µg/m³', 'count': 5770,...   

                  lastUpdated               firstUpdated  \
0   2024-09-03T22:00:00+00:00  2023-05-06T02:00:00+00:00   
1   2024-09-03T22:00:00+00:00  2016-11-18T00:00:00+00:00   
2   2024-09-03T22:00:00+00:00  2016-11-18T00:00:00+00:00   
3   2024-09-03T22:00:00+00:00  2016-11-18T00:00:00+00:00   
4   2024-09-03T22:00:00+00:00  2016-11-18T00:00:00+00:00   
5   2024-09-03T22:00:00+00:00  2016-11-18T00:00:00+00:00   
6   2024-09-03T22:00:00+00:00  2016-11-18T00:00:00

In [167]:
list(sensors_df.columns.values)

['id',
 'city',
 'name',
 'country',
 'parameters',
 'lastUpdated',
 'firstUpdated',
 'coordinates.latitude',
 'coordinates.longitude',
 'parameterIds',
 'has_pm25',
 'has_pm10',
 'has_so2',
 'has_no',
 'has_o3',
 'sensorCount']

# Get metro area coordinates

In [174]:
OW_API = 'e1fb5f641995b86dc9040b6f5ed62f33'
city_name = 'Zurich'
country_code = 'CH'

headers = {"Accept": "application/json",
           "X-API-Key": OW_API}

url = f'http://api.openweathermap.org/geo/1.0/direct?q={city_name},{country_code}&limit=10&appid={OW_API}'


In [176]:
response = requests.get(url, headers=headers)
zurich = response.json()


print(json.dumps(zurich, indent=4))

[
    {
        "name": "Zurich",
        "local_names": {
            "it": "Zurigo",
            "lv": "C\u012brihe",
            "zh": "\u8607\u9ece\u4e16",
            "is": "Z\u00fcrich",
            "br": "Z\u00fcrich",
            "gl": "Z\u00faric",
            "sk": "Z\u00fcrich",
            "rm": "Turitg",
            "de": "Z\u00fcrich",
            "fa": "\u0632\u0648\u0631\u06cc\u062e",
            "mk": "\u0426\u0438\u0440\u0438\u0445",
            "ru": "\u0426\u044e\u0440\u0438\u0445",
            "hi": "\u091c\u093c\u094d\u092f\u0942\u0930\u093f\u0916\u093c",
            "ca": "Zuric",
            "la": "Turicum",
            "ja": "\u30c1\u30e5\u30fc\u30ea\u30c3\u30d2",
            "ug": "\u0633\u064a\u06c7\u0631\u0649\u062e",
            "vo": "Z\u00fcrich",
            "eo": "Zuriko",
            "am": "\u12d9\u122a\u12ad",
            "kn": "\u0c9d\u0ccd\u0caf\u0cc2\u0cb0\u0cbf\u0c96\u0ccd",
            "id": "Z\u00fcrich",
            "my": "\u1007\u1030\u1038\u1

In [197]:
places = {}

def add_place(city, country):
    OW_API = 'e1fb5f641995b86dc9040b6f5ed62f33'
    city_name = city
    country_code = country

    headers = {"Accept": "application/json",
            "X-API-Key": OW_API}

    url = f'http://api.openweathermap.org/geo/1.0/direct?q={city_name},{country_code}&limit=1&appid={OW_API}'
    response = requests.get(url, headers=headers)
    
    metro = response.json()

    places[metro[0]['name']] = [metro[0]['lat'], metro[0]['lon']]

    return places


In [189]:
add_place('Zurich', 'CH')

{'Zurich': [47.3744489, 8.5410422]}

In [196]:
add_place('Moscow', 'RU')
add_place('Istanbul', 'TR')
add_place('London', 'GB')
add_place('Saint Petersburg', 'RU')
add_place('Berlin', 'DE')
add_place('Madrid', 'ES')
add_place('Kyiv', 'UA')
add_place('Rome', 'IT')
add_place('Baku', 'AZ')
add_place('Paris', 'FR')
add_place('Vienna', 'AT')
add_place('Minsk', 'BY')
add_place('Hamburg', 'DE')
add_place('Warsaw', 'PL')
add_place('Bucharest', 'RO')
add_place('Belgrade', 'RS')
add_place('Budapest', 'HU')
add_place('Barcelona', 'ES')
add_place('Munich', 'DE')
add_place('Kharkiv', 'UA')
add_place('Prague', 'CZ')
add_place('Milan', 'IT')
add_place('Kazan', 'RU')
add_place('Sofia', 'BG')
add_place('Tiblisi', 'GE')
add_place('Nizhny Novgorod', 'RU')
add_place('Ufa', 'RU')
add_place('Samara', 'RU')
add_place('Birmingham', 'GB')
add_place('Rostov-on-Don', 'RU')
add_place('Krasnodar', 'RU')
add_place('Yerevan', '')
add_place('Cologne', 'DE')
add_place('Voronezh', 'RU')
add_place('Perm', 'RU')
add_place('Volgograd', 'RU')
add_place('Odesa', 'UA')

{'Zurich': [47.3744489, 8.5410422],
 'Moscow': [55.7504461, 37.6174943],
 'Istanbul': [41.0091982, 28.9662187],
 'Saint Petersburg': [59.938732, 30.316229],
 'Berlin': [52.5170365, 13.3888599],
 'Kyiv': [50.4500336, 30.5241361],
 'Rome': [41.8933203, 12.4829321],
 'Baku City': [40.3755885, 49.8328009],
 'Paris': [48.8588897, 2.3200410217200766],
 'Vienna': [48.2083537, 16.3725042],
 'New York County': [40.7127281, -74.0060152],
 'London': [51.5073219, -0.1276474],
 'Madrid': [40.4167047, -3.7035825],
 'Minsk': [53.9024716, 27.5618225],
 'Hamburg': [53.550341, 10.000654],
 'Warsaw': [52.2319581, 21.0067249],
 'Bucharest': [44.4361414, 26.1027202],
 'Belgrade': [44.8178131, 20.4568974],
 'Budapest': [47.4979937, 19.0403594],
 'Barcelona': [41.3828939, 2.1774322],
 'Munich': [48.1371079, 11.5753822],
 'Kharkiv': [49.9923181, 36.2310146],
 'Prague': [50.0874654, 14.4212535],
 'Milan': [45.4641943, 9.1896346],
 'Kazan': [55.7823547, 49.1242266],
 'Sofia': [42.6977028, 23.3217359],
 'Tbilisi

In [200]:
add_place('Tirana', 'AL')
add_place('Andorra la Vella', 'AD')
add_place('Yerevan', 'AM')
add_place('Vienna', 'AT')
add_place('Baku', 'AZ')
add_place('Minsk', 'BY')
add_place('Brussels', 'BE')
add_place('Sarajevo', 'BA')
add_place('Sofia', 'BG')
add_place('Zagreb', 'HR')
add_place('Nicosia', 'CY')
add_place('Prague', 'CZ')
add_place('Copenhagen', 'DK')
add_place('Tallin', 'EE')
add_place('Helsinki', 'FI')
add_place('Paris', 'FR')
add_place('Tiblisi', 'GE')
add_place('Berlin', 'DE')
add_place('Athens', 'GR')
add_place('Budapest', 'HU')
add_place('Reykjavik', 'IS')
add_place('Dublin', 'IE')
add_place('Rome', 'IT')
add_place('Astana', 'KZ')
add_place('Riga', 'LV')
add_place('Vaduz', 'LI')
add_place('Vilnius', 'LT')
add_place('Luxembourg', 'LU')
add_place('Valletta', 'MT')
add_place('Chisinau', 'MD')
add_place('Monaco', 'MC')
add_place('Podgorica', 'ME')
add_place('Amsterdam', 'NL')
add_place('Skopje', 'MK')
add_place('Oslo', 'NO')
add_place('Warsaw', 'PL')
add_place('Lisbon', 'PT')
add_place('Bucharest', 'RO')
add_place('Moscow', 'RU')
add_place('San Marino', 'SM')
add_place('Belgrade', 'RS')
add_place('Bratislava', 'SK')
add_place('Ljubljana', 'SI')
add_place('Madrid', 'ES')
add_place('Stockholm', 'SE')
add_place('Bern', 'CH')
add_place('Ankara', 'TR')
add_place('Kyiv', 'UA')
add_place('London', 'GB')
add_place('Vatican City', 'VA')




{'Tiranë': [41.3305141, 19.825562857582966],
 'Andorra la Vella': [42.5069391, 1.5212467],
 'Yerevan': [40.1776245, 44.5126174],
 'Vienna': [48.2083537, 16.3725042],
 'Baku City': [40.3755885, 49.8328009],
 'Minsk': [53.9024716, 27.5618225],
 'Brussels': [50.8465573, 4.351697],
 'City of Sarajevo': [43.8519774, 18.3866868],
 'Sofia': [42.6977028, 23.3217359],
 'Zagreb': [45.84264135, 15.962231476593626],
 'Nicosia': [35.1748976, 33.3638568],
 'Prague': [50.0874654, 14.4212535],
 'Copenhagen': [55.6867243, 12.5700724],
 'Tallinn': [59.4372155, 24.7453688],
 'Helsinki': [60.1674881, 24.9427473],
 'Paris': [48.8588897, 2.3200410217200766],
 'Tbilisi': [41.6934591, 44.8014495],
 'Berlin': [52.5170365, 13.3888599],
 'Athens': [37.9839412, 23.7283052],
 'Budapest': [47.4979937, 19.0403594],
 'Reykjavik': [64.145981, -21.9422367],
 'Dublin': [53.3498006, -6.2602964],
 'Rome': [41.8933203, 12.4829321],
 'Astana': [51.1282205, 71.4306682],
 'Riga': [56.9493977, 24.1051846],
 'Vaduz': [47.139286

In [184]:
places = {}

print(zurich[0]['name'])
print(zurich[0]['lat'])
print(zurich[0]['lon'])

places[zurich[0]['name']] = [zurich[0]['lat'], zurich[0]['lon']]

places

Zurich
47.3744489
8.5410422


{'Zurich': [47.3744489, 8.5410422]}